internal/controller/supervisorconfig/upstreamwatcher/ldap_upstream_watcher.go

```diff
@@ -0,0 +1,367 @@
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonAuthenticationDryRunError,
+			Message: fmt.Sprintf(`failed authentication dry run for end user "%s": %s`,
+				upstream.Spec.DryRunAuthenticationUsername, err.Error()),
+		}
+	}
+
+	if !authenticated {
+		// Since we aren't doing a real auth with a password that could be wrong, the only reason we should get
+		// an unauthenticated response without an error is when the username was wrong.
+		return &v1alpha1.Condition{
+			Type:   typeLDAPConnectionValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonAuthenticationDryRunError,
+			Message: fmt.Sprintf(`failed authentication dry run for end user "%s": user not found`,
+				upstream.Spec.DryRunAuthenticationUsername),
+		}
+	}
+
+	return &v1alpha1.Condition{
+		Type:   typeLDAPConnectionValid,
+		Status: v1alpha1.ConditionTrue,
+		Reason: reasonSuccess,
+		Message: fmt.Sprintf(
+			`successful authentication dry run for end user "%s": selected username "%s" and UID "%s" [validated with Secret "%s" at version "%s"]`,
+			upstream.Spec.DryRunAuthenticationUsername,
+			authResponse.User.GetName(), authResponse.User.GetUID(),
+			upstream.Spec.Bind.SecretName, currentSecretVersion),
+	}
+}
+
+func hasPreviousSuccessfulConditionForCurrentSpecGenerationAndSecretVersion(upstream *v1alpha1.LDAPIdentityProvider, currentSecretVersion string) bool {
+	currentGeneration := upstream.Generation
+	for _, c := range upstream.Status.Conditions {
+		if c.Type == typeLDAPConnectionValid && c.Status == v1alpha1.ConditionTrue && c.ObservedGeneration == currentGeneration {
+			// Found a previously successful condition for the current spec generation.
+			// Now figure out which version of the bind Secret was used during that previous validation.
+			matches := secretVersionParser.FindStringSubmatch(c.Message)
+			if len(matches) != 2 {
+				continue
+			}
+			validatedSecretVersion := matches[1]
+			if validatedSecretVersion == currentSecretVersion {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (c *ldapWatcherController) validTLSCondition(message string) *v1alpha1.Condition {
+	return &v1alpha1.Condition{
+		Type:    typeTLSConfigurationValid,
+		Status:  v1alpha1.ConditionTrue,
+		Reason:  reasonSuccess,
+		Message: message,
+	}
+}
+
+func (c *ldapWatcherController) invalidTLSCondition(message string) *v1alpha1.Condition {
+	return &v1alpha1.Condition{
+		Type:    typeTLSConfigurationValid,
+		Status:  v1alpha1.ConditionFalse,
+		Reason:  reasonInvalidTLSConfig,
+		Message: message,
+	}
+}
+
+func (c *ldapWatcherController) validateSecret(upstream *v1alpha1.LDAPIdentityProvider, config *upstreamldap.ProviderConfig) (*v1alpha1.Condition, string) {
+	secretName := upstream.Spec.Bind.SecretName
+
+	secret, err := c.secretInformer.Lister().Secrets(upstream.Namespace).Get(secretName)
+	if err != nil {
+		return &v1alpha1.Condition{
+			Type:    typeBindSecretValid,
+			Status:  v1alpha1.ConditionFalse,
+			Reason:  reasonNotFound,
```

Comment:
This is making unsafe assumptions about the error?

Reply:
Because it could be a NotFound error or it could be some other kind of error? Maybe. Either way the resulting condition seems appropriate though.

Reply:
`reasonFailedGet`?

Reply:
The end user will see `SecretNotFound` which is appropriate for both the informer NotFound error and for any other error which is returned by the informer because either way the Secret is conceptually not found. The `Message` field will show the specific text of the error. This is also consistent with the OIDC upstream watcher controller, which is where this constant came from.